distributed/http/semaphore/prometheus.py

```diff
@@ -0,0 +1,112 @@
+import numpy as np
+
+
+class SemaphoreMetricExtension:
+    def __init__(self, dask_server):
+        self.server = dask_server
+
+    def collect(self):
+        from prometheus_client.core import (
+            GaugeMetricFamily,
+            CounterMetricFamily,
+            HistogramMetricFamily,
+        )
+        from prometheus_client.utils import INF
+
+        sem_ext = self.server.extensions["semaphores"]
+
+        semaphore_max_leases_family = GaugeMetricFamily(
+            "semaphore_max_leases",
+            "Maximum leases allowed per semaphore, this will be constant for each semaphore during its lifetime.",
+            labels=["name"],
+        )
+        for semaphore_name, max_leases in sem_ext.max_leases.items():
+            semaphore_max_leases_family.add_metric([semaphore_name], max_leases)
+        yield semaphore_max_leases_family
+
+        semaphore_active_leases_family = GaugeMetricFamily(
+            "semaphore_active_leases",
+            "Amount of currently active leases per semaphore.",
+            labels=["name"],
+        )
+        for semaphore_name, semaphore_leases in sem_ext.leases.items():
+            semaphore_active_leases_family.add_metric(
+                [semaphore_name], len(semaphore_leases)
+            )
+        yield semaphore_active_leases_family
+
+        # FIXME: Could also return a metric which shows for how long the current lease has been pending, which
+        # may be more valuable to detect leases waiting forever
+        semaphore_pending_leases = GaugeMetricFamily(
+            "semaphore_pending_leases",
+            "Amount of currently pending leases per semaphore.",
+            labels=["name"],
+        )
+        for semaphore_name, lease_ids in sem_ext.pending_leases.items():
+            semaphore_pending_leases.add_metric([semaphore_name], len(lease_ids))
+        yield semaphore_pending_leases
+
+        semaphore_acquire_total = CounterMetricFamily(
+            "semaphore_acquire_total",
+            "Total number of leases acquired per semaphore.",
+            labels=["name"],
+        )
+        for semaphore_name, count in sem_ext.metrics["acquire_total"].items():
+            semaphore_acquire_total.add_metric([semaphore_name], count)
+        yield semaphore_acquire_total
+
+        semaphore_release_total = CounterMetricFamily(
+            "semaphore_release_total",
+            "Total number of leases released per semaphore.\n"
+            "Note: if a semaphore is closed while there are still leases active, this count will not equal "
+            "`semaphore_acquired_total` after execution.",
+            labels=["name"],
+        )
+        for semaphore_name, count in sem_ext.metrics["release_total"].items():
+            semaphore_release_total.add_metric([semaphore_name], count)
+        yield semaphore_release_total
+
+        semaphore_time_to_acquire_lease = HistogramMetricFamily(
```

Comment:
I don't really have a good feeling about what this histogram values will actually look like and I would assume the size of the buckets is strongly depending on the applications running. For instance, our applications using a semaphore usually query data from a DB where these jobs usually take 10-20s or longer, including data submissions, etc. This would all fall into the max bin already. OTOH, I can see valid use cases in a much smaller range of 1-5s. For even smaller lease times I'm not even sure if this semaphore is the best fit.

I think a more stable / independent approach would be to expose some measure or quantile of the distribution. I'd start with min,max,mean,median or similar. That could be exposed in simple gauges, like `samphore_time_to_acquire_min`, etc.

Reply:
One alternative would be to change the `HistogramMetricFamily` to a `Summary` (https://prometheus.io/docs/concepts/metric_types/#summary) and then add `min`,`max`, `median`. What do you think?

Reply:
Ignore my last comment. I took a look at the prometheus_client code and there is no quantile information when using `SummaryMetricFamily` (which is the custom metric we would use instead of `Summary` in this case).
However, one can generate such quantiles with the current implementation by using https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile on the server/dashboard that is processing the metrics.

Reply:
You'd still need to define explicit buckets and we cannot do this here without very strong assumptions